### PR TITLE
expression: implement vectorized evaluation for `builtinCompressSig`

### PR DESCRIPTION
**Pull request by shihongzhi**
### What problem does this PR solve? 

implement vectorized evaluation for builtinCompressSig , for #12106

### What is changed and how it works?

	$go test -v -benchmem -bench=BenchmarkVectorizedBuiltinEncryptionFunc -run=BenchmarkVectorizedBuiltinEncryptionFunc -args "builtinCompressSig"
	goos: darwin
	goarch: amd64
	pkg: github.com/pingcap/tidb/expression
	BenchmarkVectorizedBuiltinEncryptionFunc/builtinCompressSig-VecBuiltinFunc-4         	       6	 188302054 ns/op	678877234 B/op	   19457 allocs/op
	BenchmarkVectorizedBuiltinEncryptionFunc/builtinCompressSig-NonVecBuiltinFunc-4      	       6	 198731822 ns/op	678895029 B/op	   20023 allocs/op
	PASS
	ok  	github.com/pingcap/tidb/expression	3.494s

### Check List 

Tests 

 - Unit test

### CODE REVIEW

File: expression/builtin_encryption_vec.go [hunk 1]
```diff
@@ -116,11 +118,60 @@ func (b *builtinSHA2Sig) vecEvalString(input *chunk.Chunk, result *chunk.Column)
...
+
+		resultLength := 4 + len(compressed)
+
+		// append "." if ends with space
+		shouldAppendSuffix := compressed[len(compressed)-1] == 32
+		if shouldAppendSuffix {
+			resultLength++
+		}
+
+		buffer := make([]byte, resultLength)
```
Comment (XuHuaiyu) [inline expression/builtin_encryption_vec.go]: Can we use `sync.Pool` to fetch `buffer`?
example: https://github.com/pingcap/tidb/blob/master/expression/expression.go#L216

File: expression/builtin_encryption_vec.go [hunk 2]
```diff
@@ -116,11 +118,60 @@ func (b *builtinSHA2Sig) vecEvalString(input *chunk.Chunk, result *chunk.Column)
...
+		binary.LittleEndian.PutUint32(buffer, uint32(len(str)))
+		copy(buffer[4:], compressed)
+
+		if shouldAppendSuffix {
+			buffer[len(buffer)-1] = '.'
+		}
+
+		result.AppendString(string(buffer))
```
Comment (XuHuaiyu) [inline expression/builtin_encryption_vec.go]: AppendBytes?

**Comment by qw4990** [comment]
@shihongzhi Friendly ping.

File: expression/builtin_encryption_vec.go [hunk 3]
```diff
@@ -116,11 +118,60 @@ func (b *builtinSHA2Sig) vecEvalString(input *chunk.Chunk, result *chunk.Column)
...
+		str := buf.GetString(i)
+
+		// According to doc: Empty strings are stored as empty strings.
+		if len(str) == 0 {
+			result.AppendString("")
+		}
+
+		compressed, err := deflate([]byte(str))
+		if err != nil {
+			result.AppendString(string(""))
```
[the remark this answers is not in the document]
Comment (shihongzhi) [reply expression/builtin_encryption_vec.go]: yes. according the code in `builtin_encryption.go`, there need AppendNull.
Add 'continue' keyword to continue the next loop

File: expression/builtin_encryption_vec.go (hunk 1 above)
Comment (shihongzhi) [reply expression/builtin_encryption_vec.go]: I'm already use `sync.Pool` to fetch `buffer`. pls review again

File: expression/builtin_encryption_vec.go [hunk 4]
```diff
@@ -248,11 +250,85 @@ func (b *builtinSHA2Sig) vecEvalString(input *chunk.Chunk, result *chunk.Column)
...
+		}
+
+		str := buf.GetString(i)
+
+		// According to doc: Empty strings are stored as empty strings.
+		if len(str) == 0 {
+			result.AppendString("")
+		}
+
+		compressed, err := deflate([]byte(str))
```
Comment (XuHuaiyu) [inline expression/builtin_encryption_vec.go]: s/ `[]byte(str)`/ `hack.Slice(str)`

File: expression/builtin_encryption_vec.go [hunk 5]
```diff
@@ -248,11 +250,85 @@ func (b *builtinSHA2Sig) vecEvalString(input *chunk.Chunk, result *chunk.Column)
...
+var (
+	bytePool = sync.Pool{
+		New: func() interface{} {
+			return make([]byte, defaultChunkSize)
+		},
+	}
+)
+
+func allocByteSlice(n int) []byte {
+	if n > defaultChunkSize {
```
Comment (XuHuaiyu) [inline expression/builtin_encryption_vec.go]: It not reasonable to use `defaultChunkSize` as a threshold.

Comment (shihongzhi) [reply expression/builtin_encryption_vec.go]: i use local parameter

File: expression/builtin_encryption_vec.go (hunk 4 above)
Comment (shihongzhi) [reply expression/builtin_encryption_vec.go]: sorry, i don't know what this mean. Could you give me more info

Comment (XuHuaiyu) [reply expression/builtin_encryption_vec.go]: change `[]byte(str)` to `hack.Slice(str)`

File: expression/builtin_encryption_vec.go (hunk 5 above)
Comment (XuHuaiyu) [reply expression/builtin_encryption_vec.go]: Yes I know, but `defaultChunkSize` is used for the chunk.column related pool, not for this.

File: expression/builtin_encryption_vec.go (hunk 4 above)
Comment (shihongzhi) [reply expression/builtin_encryption_vec.go]: thank you

File: expression/builtin_encryption_vec.go [hunk 6]
```diff
@@ -248,11 +251,86 @@ func (b *builtinSHA2Sig) vecEvalString(input *chunk.Chunk, result *chunk.Column)
 }
 
 func (b *builtinCompressSig) vectorized() bool {
-	return false
+	return true
+}
+
+var (
+	defaultByteSize = 1024
```
Comment (XuHuaiyu) [inline expression/builtin_encryption_vec.go]: s/ defaultByteSize/ defaultByteSliceSize

**Comment by XuHuaiyu** [comment]
PTAL @qw4990

**Pull request merged**